### PR TITLE
refactor: change runtime globals from HashSet<str> to bitflags

### PR DESCRIPTION
**Pull request by HerringtonDarkholme**
## Summary

This pull request changes runtime globals from static str to a bitflags. The reasoning here is that we don't need a HashSet to store how many flags we used. A similar trick is used previously in https://github.com/HerringtonDarkholme/vue-compiler/blob/main/crates/compiler/src/flags.rs

However, the cons is that using bitflags will only support predefined runtime helpers in Rspack. Though I think this might be a better choice since `runtime_requirements` does not accept arbitrary string.

I will add changeset and unit tests if this is a desirable change.

## Related issue (if exists)

#2452 

## Code Change 
There are mainly three changes in this pull request
1. Change `runtime_globals` from `&str` to `BitFlags`, this is the critical part. I also included some helper methods for compatiblity, e.g. converting bitflags to string_cache::Atom for swc's macro, and pretty `Display` the flags. The core logic happens in `crates/rspack_core/src/runtime_globals.rs`, for code reviewers' convenience.
3. Change `runtime_requirements` from `HashSet<&str>` to `RuntimeGlobals` in multiple structs. This involves some method name changes and borrow checking, which are all checked by compiler.
4. Rename `runtime_globals` to `RuntimeGlobals` everywhere. The change is ubiquitous but trivial.

## Perf Change

|`HashSet` usage before the change | `HashSet` usage after the change  |
|-----|----|
| image | image |

This reduces about 0.5% of HashSet overhead in compiling `arco-pro`.

## Types of changes

- [ ] Docs change / Dependency upgrade
- [ ] Bug fix
- [ ] New feature / Improvement
- [x] Refactoring
- [ ] Breaking change

## Checklist

- [ ] I have added changeset via `pnpm run changeset`.
- [ ] I have added tests to cover my changes.

### CODE REVIEW

File: crates/rspack_core/src/runtime_globals.rs [hunk 1]
```diff
@@ -1,154 +1,284 @@
...
- * runtime need to return the exports of the last entry module
- */
-pub const RETURN_EXPORTS_FROM_RUNTIME: &str = "return-exports-from-runtime";
+use std::fmt;
+
+use bitflags::bitflags;
+use swc_core::ecma::atoms::JsWordStaticSet;
+
+bitflags! {
+  pub struct RuntimeGlobals: u64 {
```
Comment (Boshen) [inline crates/rspack_core/src/runtime_globals.rs:7]: `u32`

Comment (HerringtonDarkholme) [reply crates/rspack_core/src/runtime_globals.rs:7]: I think choosing u32 will be too limiting here. If we want to add more helpers in future we will only have three slots. Adding more will introduce some unnecessary ABI change. Plus, I believe u32 perform no worse than u64 on 64bit platforms.

Comment (Boshen) [reply crates/rspack_core/src/runtime_globals.rs:7]: I just prefer to use the smallest value possible. It should be only increased when more things are added.

**Comment by underfin** [comment]
@HerringtonDarkholme Can you rebase main branch?

**Comment by HerringtonDarkholme** [comment]
> @HerringtonDarkholme Can you rebase main branch?

I have rebased the branch and pulled wasm change. Looks like we are already have 31 flags and it will probably be a growing flag set. u64 really fits here.

**Comment by Boshen** [comment]
Sorry you may have to rebase again, I just merged something :-(

**Comment by Boshen** [comment]
@underfin merge?

**Review by Boshen — APPROVED** [review]

**Review by underfin — APPROVED** [review]

**Comment by underfin** [comment]
Thanks a lot!

**Pull request merged**